sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1405,6 +1405,27 @@ object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
 }
 
 /**
+  * Replaces logical [[Except]] operator with a left-anti [[Join]] operator.
+  * {{{
+  *   SELECT a1, a2 FROM Tab1 EXCEPT SELECT b1, b2 FROM Tab2
+  *   ==>  SELECT DISTINCT a1, a2 FROM Tab1 LEFT ANTI JOIN Tab2 ON a1<=>b1 AND a2<=>b2
+  * }}}
+  *
+  * Note:
+  * 1. This rule is only applicable to EXCEPT DISTINCT. Do not use it for INTERSECT ALL.
+  * 2. This rule has to be done after de-duplicating the attributes; otherwise, the generated
+  *    join conditions will be incorrect.
+  */
+object ReplaceIntersectWithAntiJoin extends Rule[LogicalPlan] {
```

Comment:
We could also integrate this with `ReplaceIntersectWithSemiJoin`. But I'll leave that up to you.